sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,518 @@
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An unbounded source for MQTT broker.
+ *
+ * <h3>Reading from a MQTT broker</h3>
+ *
+ * <p>MqttIO source returns an unbounded collection of {@code byte[]} as
+ * {@code PCollection<byte[]>}, where {@code byte[]} is the MQTT message payload.</p>
+ *
+ * <p>To configure a MQTT source, you have to provide a MQTT connection configuration including
+ * {@code ClientId}, a {@code ServerURI}, and eventually a {@code Topic} pattern. The following
+ * example illustrates various options for configuring the source:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(
+ *   MqttIO.read()
+ *    .withMqttConnectionConfiguration(MqttIO.MqttConnectionConfiguration.create(
+ *      "tcp://host:11883",
+ *      "my_client_id",
+ *      "my_topic"))
+ *
+ * }</pre>
+ *
+ * <h3>Writing to a MQTT broker</h3>
+ *
+ * <p>MqttIO sink supports writing {@code byte[]} to a topic on a MQTT broker.</p>
+ *
+ * <p>To configure a MQTT sink, as for the read, you have to specify a MQTT connection
+ * configuration with {@code ClientId}, {@code ServerURI}, {@code Topic}.</p>
+ *
+ * <p>Eventually, you can also specify the {@code Retained} and {@code QoS} of the MQTT
+ * message.</p>
+ *
+ * <p>For instance:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline
+ *   .apply(...) // provide PCollection<byte[]>
+ *   .MqttIO.write()
+ *     .withMqttConnectionConfiguration(MqttIO.MqttConnectionConfiguration.create(
+ *       "tcp://host:11883",
+ *       "my_client_id",
+ *       "my_topic"))
+ *     .withRetained(true)
+ *     .withQoS(2)
+ *
+ * }</pre>
+ */
+public class MqttIO {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MqttIO.class);
+
+  public static Read read() {
+    return new AutoValue_MqttIO_Read.Builder()
+        .setMaxReadTime(null).setMaxNumRecords(Long.MAX_VALUE).build();
+  }
+
+  public static Write write() {
+    return new AutoValue_MqttIO_Write.Builder()
+        .setRetained(false)
+        .setQos(0)
+        .build();
+  }
+
+  private MqttIO() {
+  }
+
+  /**
+   * A POJO describing a MQTT connection.
+   */
+  @AutoValue
+  abstract static class MqttConnectionConfiguration implements Serializable {
+
+    @Nullable abstract String serverUri();
+    @Nullable abstract String clientId();
+    @Nullable abstract String topic();
+
+    public static MqttConnectionConfiguration create(String serverUri, String clientId,
+                                                     String topic) {
+      checkNotNull(serverUri, "serverUri");
+      checkNotNull(clientId, "clientId");
+      checkNotNull(topic, "topic");
+      return new AutoValue_MqttIO_MqttConnectionConfiguration(serverUri, clientId, topic);
+    }
+
+    private void validate() {
```

Comment:
Isn't this redundant with the validation happening in create()? - create() should not allow to create an invalid object.